tests/src/app/testqgsmaptoolselect.cpp

```diff
@@ -0,0 +1,154 @@
+    {}
+
+  private slots:
+    void initTestCase(); // will be called before the first testfunction is executed.
+    void cleanupTestCase(); // will be called after the last testfunction was executed.
+    void init(); // will be called before each testfunction is executed.
+    void cleanup(); // will be called after every testfunction.
+
+    void selectInvalidPolygons(); // test selecting invalid polygons
+
+  private:
+    QgsMapCanvas* canvas;
+
+    QgsFeatureList testSelectVector( QgsVectorLayer* layer, double xGeoref, double yGeoref );
+
+    // Release return with delete []
+    unsigned char *
+    hex2bytes( const char *hex, int *size )
+    {
+      QByteArray ba = QByteArray::fromHex( hex );
+      unsigned char *out = new unsigned char[ba.size()];
+      memcpy( out, ba.data(), ba.size() );
+      *size = ba.size();
+      return out;
+    }
+
+    // TODO: make this a QgsGeometry member...
+    QgsGeometry geomFromHexWKB( const char *hexwkb )
+    {
+      int wkbsize;
+      unsigned char *wkb = hex2bytes( hexwkb, &wkbsize );
+      QgsGeometry geom;
+      // NOTE: QgsGeometry takes ownership of wkb
+      geom.fromWkb( wkb, wkbsize );
+      return geom;
+    }
+
+};
+
+void TestQgsMapToolSelect::initTestCase()
+{
+  QgsApplication::init();
+  QgsApplication::initQgis();
+  // Set up the QSettings environment
+  QCoreApplication::setOrganizationName( "QGIS" );
+  QCoreApplication::setOrganizationDomain( "qgis.org" );
+  QCoreApplication::setApplicationName( "QGIS-TEST" );
+
+  QgsApplication::showSettings();
+
+  // enforce C locale because the tests expect it
+  // (decimal separators / thousand separators)
+  QLocale::setDefault( QLocale::c() );
+}
+
+void TestQgsMapToolSelect::cleanupTestCase()
+{
+  QgsApplication::exitQgis();
+}
+
+void TestQgsMapToolSelect::init()
+{
+  canvas = new QgsMapCanvas();
+}
+
+void TestQgsMapToolSelect::cleanup()
+{
+  delete canvas;
+}
+
+// private
+QgsFeatureList
+TestQgsMapToolSelect::testSelectVector( QgsVectorLayer* layer, double xGeoref, double yGeoref )
+{
+  QScopedPointer< QgsMapToolSelect > tool( new QgsMapToolSelect( canvas ) );
+  QgsPoint mapPoint = canvas->getCoordinateTransform()->transform( xGeoref, yGeoref );
+
+  // make given vector layer current
+  canvas->setCurrentLayer( layer );
+
+  QScopedPointer< QgsMapMouseEvent > event( new QgsMapMouseEvent(
+        canvas,
+        QEvent::MouseButtonRelease,
+        QPoint( mapPoint.x(), mapPoint.y() )
+      ) );
+
+  // trigger mouseRelease handler
+  tool->canvasReleaseEvent( event.data() );
+
+  // return selected features
+  return layer->selectedFeatures();
+}
+
+void TestQgsMapToolSelect::selectInvalidPolygons()
+{
+  //create a temporary layer
+  QScopedPointer< QgsVectorLayer > memoryLayer( new QgsVectorLayer( "Polygon?field=pk:int", "vl", "memory" ) );
+  QVERIFY( memoryLayer->isValid() );
+  QgsFeature f1( memoryLayer->dataProvider()->fields(), 1 );
+  f1.setAttribute( "pk", 1 );
+  f1.setGeometry( geomFromHexWKB(
```

Comment:
Mind adding a comment with the expected vertices for this wkb here? It'll make the test more readable in future.